arbor/partition_load_balance.cpp

```diff
@@ -222,5 +222,110 @@ domain_decomposition partition_load_balance(
     return d;
 }
 
+domain_decomposition partition_by_group(const recipe& rec, const context& ctx, const std::vector<group_description>& groups) {
+    struct partition_gid_domain {
+        partition_gid_domain(const gathered_vector<cell_gid_type>& divs, unsigned domains) {
+            auto rank_part = util::partition_view(divs.partition());
+            for (auto rank: count_along(rank_part)) {
+                for (auto gid: util::subrange_view(divs.values(), rank_part[rank])) {
+                    gid_map[gid] = rank;
+                }
+            }
+        }
+        int operator()(cell_gid_type gid) const {
+            return gid_map.at(gid);
+        }
+        std::unordered_map<cell_gid_type, int> gid_map;
+    };
+
+    unsigned num_domains = ctx->distributed->size();
+    unsigned domain_id = ctx->distributed->id();
+    auto num_global_cells = rec.num_cells();
+
+    std::vector<cell_gid_type> local_gids;
+    for (const auto& g: groups) {
+        local_gids.insert(local_gids.end(), g.gids.begin(), g.gids.end());
+    }
+    cell_size_type num_local_cells = local_gids.size();
+
+    auto global_gids = ctx->distributed->gather_gids(local_gids);
+
+    domain_decomposition d;
```

Comment:
This seems quite old-school (Cxx20 will gives us dedicated initializers, but still).
Does this look better to you?
```
  return { num_domains, domain_id, 
           num_local_cells, num_global_cells, 
           groups, partition_gid_domain(global_gids, num_domains)};
```

Reply:
I think I like the C++20 way and right now both variants are terrible ;)